### ISSUE TITLE
pgsql: can't determine url id type

### ISSUE THREAD

**Issue by amaurybrisou** [strongloop/loopback#1880]
Hello guys, 

I'm currenctly stuck in a ManytoMany self-referencing with through table, on one step only.

I have this main model self referencing itself :

```javascript
{
  "name": "Surfer",
  "base": "User",
  "idInjection": true,
  "options": {
    "validateUpsert": true,
    "postgresql": {
      "schema": "couch-surfing",
      "table": "surfer"
    }
  },
  "properties": {
    "birthdate": {
      "type": "date"
    },
    "telephone": {
      "type": "string"
    }
  },
  "validations": [],
  "relations": {
    "friends": {
      "type": "hasMany",
      "model": "Surfer",
      "foreignKey":"surfer",
      "through": "SurferFriend",
      "keyThrough": "friend",
    }
  },
  "acls": [],
  "methods": {}
}

```

then the through table : 

```javascript
{
  "name": "SurferFriend",
  "base": "PersistedModel",
  "idInjection": true,
  "options": {
    "validateUpsert": true,
    "postgresql": {
      "schema": "couch-surfing",
      "table": "surfer-friend"
    }
  },
  "properties": {
    "from": {
      "type": "date",
      "required": true,
      "defaultFn": "now"
    },
    "surfer": {
      "type": "number",
      "required": true
    },
    "friend": {
      "unique": true,
      "type": "number",
      "required": true
    }
  },
  "validations": [],
  "relations": {},
  "acls": [],
  "methods": {}
}
```

Everything works great for fetching friends, If I add manually a SurferFriend relation with two Ids in the post body, along with the other fields of course; It works !
My problem comes when I use the explorer suggested url : 

PUT /Surfers/{id}/friends/{fk}                    Update a related item by id for friends.

I use : 

PUT http://localhost:3000/api/Surfer/1/friends/2 

post data : 
```json 
   { 
       "from": "Mon Dec 14 2015 01:00:00 GMT+0100 (CET)",
       "surfer": 1,
       "friend": 2 
   }
```
it returns me this : 

    strong-remoting:shared-method - __link__friends - invoke with +0ms [ 2,
    { from: Mon Dec 14 2015 01:00:00 GMT+0100 (CET),
    surfer: 1,
    friend: 2 },
    [Function: callback] ]
    { type: [Function: Number], id: true, required: true } 1
    undefined 2
    strong-remoting:shared-method error caught during the invocation of __link__friends +8ms
     express:router restRemoteMethodNotFound  : /api/Surfers/1/friends/rel/2?access_token=C3t3PEbcFtA8U1gsAN90ir8dCy6njjrTs2TzEpQo9BtmHoQt1TsFJ0pz1q1mDRVU +0ms
    express:router restUrlNotFound  : /api/Surfers/1/friends/rel/2?access_token=C3t3PEbcFtA8U1gsAN90ir8dCy6njjrTs2TzEpQo9BtmHoQt1TsFJ0pz1q1mDRVU +5ms
    express:router restErrorHandler  : /api/Surfers/1/friends/rel/2?access_token=C3t3PEbcFtA8U1gsAN90ir8dCy6njjrTs2TzEpQo9BtmHoQt1TsFJ0pz1q1mDRVU +1ms
    strong-remoting:rest-adapter Error in PUT /Surfers/1/friends/rel/2?access_token=C3t3PEbcFtA8U1gsAN90ir8dCy6njjrTs2TzEpQo9BtmHoQt1TsFJ0pz1q1mDRVU:   TypeError: Cannot read property 'type' of undefined
     at PostgreSQL.toColumnValue (/home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-connector-postgresql/lib/postgresql.js:433:11)
     at PostgreSQL.SQLConnector._buildWhere (/home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-connector-postgresql/node_modules/loopback-connector/lib/sql.js:813:26)
     at PostgreSQL.SQLConnector.buildWhere (/home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-connector-postgresql/node_modules/loopback-connector/lib/sql.js:638:26)
     at PostgreSQL.SQLConnector.buildSelect (/home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-connector-postgresql/node_modules/loopback-connector/lib/sql.js:1001:28)
     at PostgreSQL.find [as all] (/home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-connector-postgresql/node_modules/loopback-connector/lib/sql.js:1058:19)
     at /home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-datasource-juggler/lib/dao.js:1511:42
     at doNotify (/home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
     at doNotify (/home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
     at doNotify (/home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
     at doNotify (/home/erol/nodejs/drafts/CouchSporting/node_modules/loopback-datasource-juggler/lib/observer.js:93:49) +2ms

    ps: the 'undefined 2' line is my personal log I added into the function toColumnValue for debug purpose

I don't really see what specific option I could give to loopback configurations to let it work as expected! 

Thank you for you reads and patience !

Amaury Brisou

**Comment by loay** [comment]
Hi @amaurybrisou 
Thanks for reporting the issue?
Is your repository available here so I can have a look?

**Comment by amaurybrisou** [comment]
Hi sorry for the long delay, 

I've found the solution and the actual documentation of hasManySelfThrough is right as it is. I just missunderstood at first. 

solution :  add the two belongsTo Surfer on the SurferFriend model with foreignKey friendId and userId. these Keys have to be declared in the SurferFriend properties !

On the other side, Surfer, I added the inverted relation friendsOf but it's not mandatory. 

Note: creating a friend doesn't create friendsOf, of course you'd tell me, but it's annoying for an ORM.

Thanks :)
